yoyodyne/evaluators.py

```diff
@@ -155,3 +221,130 @@ def finalize_predictions(
             with torch.inference_mode():
                 predictions[i] = torch.cat((symbols, pads))
         return predictions
+
+    def finalize_golds(
+        self,
+        golds: torch.Tensor,
+        *args,
+        **kwargs,
+    ):
+        return golds
+
+    @property
+    def metric_name(self):
+        return "accuracy"
+
+
+class CEREvaluator(Evaluator):
+    """Evaluates character error rate."""
+
+    def get_eval_item(
+        self,
+        predictions: List[List[str]],
+        golds: List[List[str]],
+        pad_idx: int,
+    ) -> EvalItem:
+        cer_calc = CharErrorRate()
+        cers = []
```

Comment:
Good point. I think I did it this way b/c I think its what the `torchmetrics.CharErrorRate` seemed to expect. I can check their docs and try to avoid this.

Reply:
Okay, actually I realize there is a semantic ambiguity here. I think we could have the following, in theory:

* "symbol error rate" or SER which is done at the tensor level
* actually join the targets using `--target_sep` (which has to be passed in for correctness) and then computing true CER

I suspect when there's a difference most people want the former.

Reply:
Right! This is what I meant with:

> it is essentially whatever-you-define-one-symbol-to-be error rate, which in my case happens to be character (or phoneme).

In the second bullet, the joining entails decoding into strings right? In that case I agree. We **could** have an option for both, though the naming could get confusing... E.g. I might ask for cer because I know what it is, not realizing that ser will be the same, but computed faster.

I do like the idea of getting metrics faster by leaving them as tensors, but I think there could be many such cases where we might want to evaluate against strings, and for a developer, adding a new evaluator is probably easier if you can get strings rather than figuring out the right tensor logic. Maybe this is why e.g. huggingface seems to decode into strings by default.

Another example eval that I need to implement is transliteration, where typically for one word you have several valid targets. Eval can then compute the number of top-k hypothesis that are in the list of targets, or do some fuzzy matching, or ranking, etc. This could all be fine in that a particular evaluator could just decode into strings if need be, but then we have to potentially decoding the same preds into strings many times, and maybe the same golds as well!

Reply:
Did you mean to edit my comment haha?

Reply:
Oh god no sorry I meant to quote-reply to it with that.

Reply:
Just gonna repost the original down here:

Right! This is what I meant with:

it is essentially whatever-you-define-one-symbol-to-be error rate, which in my case happens to be character (or phoneme).

In the second bullet, the joining entails decoding into strings right? In that case I agree. We could have an option for both, though the naming could get confusing... E.g. I might ask for cer because I know what it is, not realizing that ser will be the same, but computed faster.

I do like the idea of getting metrics faster by leaving them as tensors, but I think there could be many such cases where we might want to evaluate against strings, and for a developer, adding a new evaluator is probably easier if you can get strings rather than figuring out the right tensor logic. Maybe this is why e.g. huggingface seems to decode into strings by default.

Another example eval that I need to implement is transliteration, where typically for one word you have several valid targets. Eval can then compute the number of top-k hypothesis that are in the list of targets, or do some fuzzy matching, or ranking, etc. This could all be fine in that a particular evaluator could just decode into strings if need be, but then we have to potentially decoding the same preds into strings many times, and maybe the same golds as well!